readme.md

```diff
@@ -15,834 +14,1113 @@
+	if ( foo ) {
+		return 'foo';
+	}
+	return 'quux';
+}
 
-    ```javascript
+// Bad:
+function doSthComplex( foo, bar ) {
+	if ( foo ) {
+		// do something
 
-    // 6.B.1
-    function Selection() {
-        var editor = this.editor;
+		if ( bar ) {
+			// do something
+		} else {
+			// do something else
+		}
+	}
+}
 
-        // Bad:
-        editor.on( 'setData', someCallback, editor );
+// Good:
+function doSthComplex( foo, bar ) {
+	if ( !foo )
+		return;
 
-        // Still bad:
-        editor.on( 'setData', CKEDITOR.tools.bind( someCallback, editor ) );
+	// do something
 
-        // Good:
-        editor.on( 'setData', someCallback, this );
+	if ( !bar ) {
+		// do something else
+		return;
+	}
 
-        // Good:
-        this.getEditor().on( 'setData', function( evt ) {
-            evt.editor.doSomething();
-        } );
+	// do something
+}
+```
 
-        // Good too:
-        var that = this;
-        editor.on( 'setData', function( evt ) {
-            editor.setCurrentSelection( that );
-        } );
-    }
+### 2.8.2. Objects minify poorly, closures very well
 
-    function someCallback() {
-        // Should refers to Selection, not Editor.
-        this;
-    }
+Avoid excessing usage of objects, when functional programming may be used as well. Properties and method names unlike private variables and function names cannot be shortened by minifier.
+You can play with [online Google Closure Compiler](http://closure-compiler.appspot.com/home) to test different approaches. Make sure to check the gzipped size which matters the most.
 
-    ```
 
-    C. Use `thisArg`
+## 2.9. Native & Host Objects
 
-    Several prototype methods of ES 5.1 built-ins come with a special `thisArg` signature, which should be used whenever possible. CKEditor follows this practice too, so for example [`eventTarget.on`](http://docs.ckeditor.com/#!/api/CKEDITOR.event-method-on) also accepts context as an argument.
+The basic principle here is:
 
-    ```javascript
+	Don't do stupid shit and everything will be ok.
 
-    // 6.C.1
+To reinforce this concept, please watch the following presentation:
+	
+[“Everything is Permitted: Extending Built-ins” by Andrew Dupont (JSConf2011, Portland, Oregon)](https://www.youtube.com/watch?v=xL3xCO7CLNM)
 
-    var obj = { f: 'foo', b: 'bar', q: 'qux' };
+**But** CKEditor is a widget - an application inside someone's system &ndash; so... do not touch anything except
+`CKEDITOR.*` namespace.
 
-    Object.keys( obj ).forEach( function( key ) {
 
-      // |this| now refers to `obj`
+## 2.10. Comments
+- Single line above the code that is subject (except for long expressions like conditions).
+- Comment is a sentence - start it with capital leter, end with period.
+- Use multiline comments only for API documentation.
+- Check the [JSDuck documentation guide](http://dev.ckeditor.com/wiki/CodeDocumentation).
+- You can use JSDuck documentation style for private functions, but then use single line comments.
```

Comment:
Actually we often use multiline, but non-JSDocs comments:

```javascript
/*
 * Some docs for private functions.
 */
```

Reply:
I believe it's more for helper functions where we use `//` instead to avoid producing incorrect JSDocs :thinking:

Reply:
Please consider comments like this one:
https://github.com/ckeditor/ckeditor4/blob/294ea3333aa3d2b796f27ca8503ac7c1a658d50a/plugins/basicstyles/plugin.js#L98-L102

Reply:
@msamsel it rather seems like this example is incorrect because array could be split across multiple lines instead of using multiline comment :thinking: WDYT?

Reply:
yeah maybe it's not perfect then :D